Expose the getResourcePath and getRootResourcePath wrappers.

cyphar opened the pull request:
Due to the importance of path safety, the internal sanitisation wrappers
for volumes and containers should be exposed so other parts of Docker
can benefit from proper path sanitisation.

Signed-off-by: Aleksa Sarai cyphar@cyphar.com (github: cyphar)

cyphar committed:
*: expose getResourcePath and getRootResourcePath wrappers

Due to the importance of path safety, the internal sanitisation wrappers
for volumes and containers should be exposed so other parts of Docker
can benefit from proper path sanitisation.

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com> (github: cyphar)

cyphar commented:
/cc @cpuguy83 @shykes @vieux @crosbymichael @unclejack @tiborvass @vishh

tiborvass commented:
LGTM

duglin commented:
LGTM

LK4D4 commented:
@cyphar Is there any particular places? Because I think that both `container` and `volume` can be fully private.

cyphar commented:
@LK4D4 There's quite a few places (see: the builder) where they use their own path sanitisation rather than just using these wrappers.

cpuguy83 commented:
@cyphar Where?
I wouldn't expect anything outside the stuff in the daemon package to be looking into the container like that.

cyphar commented:
`builder/internals.go` inside `addContext` is one example. And I think there's somewhere else in the builder too.

cpuguy83 commented:
:( This should be a call on the daemon.

LK4D4 commented:
@cyphar Then you should use them there. I'm not in favor having not used public methods (I know that there are already a lot of them :))

cyphar committed:
*: switch to Get(Root)?ResourcePath where appropriate

Several parts of the codebase didn't use the correct path sanitisation
wrappers. Now that the wrappers have been exposed, use those.

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com> (github: cyphar)

cyphar commented:
@LK4D4 @cpuguy83 I've fixed up those cases to use the newly exposed wrappers (there were also two cases where we didn't use `get(Root)?ResourcePath` inside `daemon`). Also rebase'd.

LK4D4 commented:
@cyphar Something pretty bad with tests. Are they passing on your machine?

aluzzardi commented:
ping @cyphar

cyphar commented:
Yes, sorry. The second commit's to blame -- I will fix it this weekend.

stevvooe commented on daemon/container.go:
```diff
@@ -204,12 +204,12 @@ func (container *Container) LogEvent(action string) {
 }
 
-func (container *Container) getResourcePath(path string) (string, error) {
+func (container *Container) GetResourcePath(path string) (string, error) {
```
If these are exported, they should have godoc comment explaining exactly what is provided.

stevvooe commented:
LGTM, except for documentation about what "resource path" and a "root resource path" actually are.

Thank you for the nice refactoring!

cyphar commented:
There, I fixed the problem (`container.RootfsPath` is a weird name for something that returns the `basefs`, given that `root` has a different meaning, btw). PTAL.

/cc @cpuguy83 @shykes @vieux @crosbymichael @unclejack @tiborvass @vishh

stevvooe commented on volumes/volume.go:
```diff
@@ -146,14 +146,14 @@ func (v *Volume) FromDisk() error {
-	return v.getRootResourcePath("config.json")
+	return v.GetRootResourcePath("config.json")
 }
-func (v *Volume) getRootResourcePath(path string) (string, error) {
```
Did these need to be documented, as well?

tiborvass replied:
@cyphar would you mind adding just a small documentation for GetRootResourcePath and GetResourcePath, even if it's just "go look at contianer's doc"

cyphar replied:
```
// Don't use this if you don't know what you're doing. ;]
```

But yeah, will do.

cyphar replied:
Fix'd.
/ping @tiborvass

tiborvass commented on daemon/container.go:
```diff
@@ -208,12 +208,37 @@ func (container *Container) LogEvent(action string) {
 }
 
-func (container *Container) getResourcePath(path string) (string, error) {
+// Evaluates `path` in the scope of the container's basefs, with proper path
```
in the code it's mostly called rootfs though so idk :S

cyphar commented:
There, I've fixed up the comments. PTAL.

/ping @cpuguy83 @shykes @vieux @crosbymichael @unclejack @tiborvass @vishh @stevvooe

tiborvass commented:
LGTM

tiborvass commented:
oh gosh, @cyphar would you mind rebasing? Sorry :(

cyphar commented:
`rebase`'d. pls2merge kthxbai.

/ping @cpuguy83 @shykes @vieux @crosbymichael @unclejack @tiborvass @vishh @stevvooe

cyphar commented:
rebase'd. pls2merge kthxbai.

/ping @cpuguy83 @shykes @vieux @crosbymichael @unclejack @tiborvass @vishh @stevvooe

cpuguy83 commented:
LGTM
